RUM libraries specification (#112)
signalfx/gdi-specification: specification/configuration.md

configuration. An advanced installation process can be provided where the user
chooses what to install limiting the configuration options.

> ## Configuration options

Author: The title/main header is already named `Configuration`, I think it's slightly better to specify that this section describes just supported config options.

  - Distribution MUST default to `otlp`
  - Distribution MUST offer `jaeger-thrift-splunk` that defaults to `http://127.0.0.1:9080/v1/trace`

### Real User Monitoring Libraries

Real User Monitoring (RUM) instrumentation libraries cannot use environment variables for configuration.
Instead, they MUST expose a `SplunkRum` object/class/namespace (depending on the language used) that allows
setting the following properties (in a language-specific way: Java may use builders, Swift can use named
parameters with default values, JavaScript can use objects, etc.):

| Configuration property (default value) | Description                                                                                                                                                                                                 |
| -------------------------------------- | -----------                                                                                                                                                                                                 |

> | `beaconUrl` ()                         | RUM beacon URL, ex. `https://rum-ingest.<realm>.signalfx.com/v1/rum`. [1]                                                                                                                                   |

Reviewer: Implementations MUST enforce by default that beaconUrls are `https` only, and reject/fail to start otherwise.  Implementations MAY offer an `allowInsecureBeacon` option (default `false`) for customers wishing to use that.

Reviewer 2: Does this need to be done by the library, though? Android does this by default, so I don't know that we should be doing it on top of the android way it works.

Reviewer: The outcome should ideally be something the app developer can notice (an exception throw, error code returned, or error printout visible without additional config) the first time they integrate.  If Android is silently failing to send the wire traffic I'd bother to add the extra 3 lines of if/print/throw.

Reviewer 2: No, it's not silent. Android will crash the app if you try it. :)

Author: The outcome is the same (app is failing), does it really matter if the check is built-in in the platform or if we do it ourselves?

> | `rumAccessToken` ()                    | RUM authentication token. [1]                                                                                                                                                                               |

Reviewer: We use the term `rumAuth` for this.

Author: Yes, but should we continue to use `rumAuth`? `rumAccessToken` is more similar to `SPLUNK_ACCESS_TOKEN` that is used in collector & instrumentation libs, and it is an "access token" (I mean you can find it on the "Access Tokens" page in the UI)

Reviewer: It is named differently because it is different in kind.  `rumAuth` is public and we don't want anybody pasting a random other access token into that space.

Author: Fair point -- I'll revert to `rumAuth`.

> | `applicationName` ()                   | Instrumented application name. [1]                                                                                                                                                                          |

Reviewer: We use the term `app` for this.

Reviewer 2: on the programmatic configuration? not `appName` ?

Reviewer: Yep - https://github.com/signalfx/splunk-otel-js-web/blob/main/docs/Configuration.md#general-configuration-options - not built in iOS yet but let's be consistent.  In general I like it when the config name matches up with the attribute name.

Reviewer 2: hmm. can't say that I'm a fan of cryptic configuration method names.

Author: Connected to https://github.com/signalfx/gdi-specification/pull/112#discussion_r657296227: should we continue to use `app`/`appName`/`applicationName`? `serviceName` & `service.name` would be more consistent with otel conventions

Reviewer 2: Like @flands said, the user should only have to set one thing (IMO the application name, whatever we call that configuration item), and the library should do the needful under the hood and set both `app` on spans, and `service.name` on the Resource to that value.

| `globalAttributes` ()                  | OpenTelemetry [Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes) that will be added to every span produced by the RUM library. |

> 

Reviewer: We should also enforce that an `environmnt` config option is available.

Reviewer 2: yes, I forgot this was something that we should add to the android library.

Author: 👍 

Should we use `deploymentEnvironment` option (and `deployment.environment` attribute) to match otel conventions?

Currently just `environment` is used: https://github.com/signalfx/splunk-otel-js-web/blob/main/src/index.ts#L265

Reviewer 3: Alright, but if something is clearly just an attribute (that has a semantic meaning in our backend, but doesn't have a semantic meaning in the instrumentation layer), then shouldn't we just say "add it to the global attributes"? It sort of applies similarly to application name.

Author: I think there's value in adding separate config options for the most commonly used attributes - it eases the setup.

We're already doing that in otel sdk: there's now `OTEL_SERVICE_NAME=...` option supported even though you could use `OTEL_RESOURCE_ATTRIBUTES=service.name=...` to define exactly the same thing.

> that has a semantic meaning in our backend, but doesn't have a semantic meaning in the instrumentation layer

It has meaning for RUM as a whole though

- [1] These configuration properties are mandatory and MUST be provided by the user. If any of these is missing,
  the RUM instrumentation library MUST fail to start.

Other requirements:

> - RUM library MUST use the Zipkin span exporter by default

Author: I guess that we have a hard dependency on our RUM backend here; is this the desired state though? We could use otel resources if we didn't use zipkin and used otlp json instead

Reviewer 2: For the moment, it's the only way to get data into the RUM backend, so I think it can be a MUST for now (otherwise nothing will work)

  - Distribution MUST default to `otlp`
  - Distribution MUST offer `jaeger-thrift-splunk` that defaults to `http://127.0.0.1:9080/v1/trace`

### Real User Monitoring Libraries

Real User Monitoring (RUM) instrumentation libraries cannot use environment variables for configuration.
Instead, they MUST expose a `SplunkRum` object/class/namespace (depending on the language used) that allows
setting the following properties (in a language-specific way: Java may use builders, Swift can use named
parameters with default values, JavaScript can use objects, etc.):

| Configuration property (default value) | Description                                                                                                                                                                                                 | Required |
| -------------------------------------- | -----------                                                                                                                                                                                                 | -------- |

> | `beaconUrl` ()                         | RUM beacon URL, ex. `https://rum-ingest.<realm>.signalfx.com/v1/rum`. [1] [2]                                                                                                                               | Yes      |

Reviewer 3: 1) Is there a reason this doesn't default to us0 endpoint? 2) Could we have an alternative `beaconEndpoint = 'us0' | 'us1' | ...`, where if you set `beaconEndpoint` then you don't have to set `beaconUrl`? `beaconUrl` would still be available.

Author: It's a good idea, the collector already does something similar with realm/ingest URL. I'll add it, thanks!

| `rumAccessToken` ()                    | RUM authentication token. [1]                                                                                                                                                                               | Yes      |
| `applicationName` ()                   | Instrumented application name. [1]                                                                                                                                                                          | Yes      |
| `globalAttributes` ()                  | OpenTelemetry [Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes) that will be added to every span produced by the RUM library. | Yes      |

> | `allowInsecureBeacon` (false)          | Allows sending data to insecure endpoints not using https. [2]                                                                                                                                              | No       |

Reviewer 3: Since we're codifying this as a spec now, I don't see a reason for this parameter at all in Browser RUM. This problem is already solved by browser vendors https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content . And provided that our backend rejects insecure incoming traffic (I don't mean redirecting to HTTPS, just straight rejection), this shouldn't be an issue for mobile RUMs either.

Author: It doesn't make sense for Android too, since it rejects all non-https connections by default. @johnbley is that option only for iOS?

Reviewer: It's not just about ensuring that any traffic actually sent is https, it's about ensuring that developers very quickly know that they have mistyped something, rather than silently dropping data on the ground or allowing data to bleed out in plaintext across the internet only to be dropped by our ingest.  I would only believe it's not necessary in browser with demonstrable proof that browsers do the sane thing.  For android does it loudly and immediately complain or does it only complain later and only if you turn on debug logging?  For iOS this can be misconfigured, so, yes, it is necessary there.  It is certainly the case that we can simply not support the idea of `allowInsecureBeacon` on any given platform (particularly since some may make that impossible), but we **should** have active initialization-time checking that we are using https.

Reviewer 2: As mentioned above, Android will crash your app if you try to use an insecure endpoint without specifically adding the additional permission to your app manifest.

Reviewer 4: Another option is to keep it as a developer flag -- do not add to README but to CONTRIBUTING and make it clear it is unsupported and may be removed.

Author: Do we need to keep it in the spec if it's just a developer flag? If it's just a dev setting just for the iOS library then we can probably remove it from the spec.

  - Distribution MUST default to `otlp`
  - Distribution MUST offer `jaeger-thrift-splunk` that defaults to `http://127.0.0.1:9080/v1/trace`

> ### Real User Monitoring Libraries

Reviewer 4: nit: either this should be indented under Instrumentation Libraries or Instrumentation Libraries should further define what is in scope -- I assume it is applications and serverless.

Author: Good point - I'll put that under the Istrumentation Libraries section. Thanks!

Real User Monitoring (RUM) instrumentation libraries cannot use environment variables for configuration.
Instead, they MUST expose a `SplunkRum` object/class/namespace (depending on the language used) that allows
setting the following properties (in a language-specific way: Java may use builders, Swift can use named
parameters with default values, JavaScript can use objects, etc.):

> | Configuration property (default value) | Description                                                                                                                                                                                                 | Required |

Reviewer 4: nit: We are already in the Configuration section -- can we shorten the name to just Property?

```suggestion
| Property (default value) | Description                                                                                                                                                                                                 | Required |
```

| -------------------------------------- | -----------                                                                                                                                                                                                 | -------- |

> | `realm` ()                             | Splunk realm, e.g. `us0`, `us1`. If set, value of `beaconUrl` will be automatically computed based on this. [1] [2]                                                                                         | Yes      |

Reviewer 4: Just to confirm, we are naming this `realm` and not something like `splunkRealm` because we are in the `SplunkRum` namespace?

Author: Yeah, that was exactly what I thought - "Splunk" is already in `SplunkRum`, no need to repeat it.

> | `realm` ()                             | Splunk realm, e.g. `us0`, `us1`. If set, value of `beaconUrl` will be automatically computed based on this. [1] [2]                                                                                         | Yes      |

Reviewer 4: I think the [1] should move next to the Yes for required

```suggestion
| `realm` ()                             | Splunk realm, e.g. `us0`, `us1`. If set, value of `beaconUrl` will be automatically computed based on this. [2]                                                                                         | Yes [1]      |
```

Author: No longer applicable

> | `beaconUrl` ()                         | RUM beacon URL, ex. `https://rum-ingest.<realm>.signalfx.com/v1/rum`. If both `realm` and `beaconUrl` are set, `beaconUrl` takes precedence. [1] [2]                                                        | Yes      |

Reviewer 4: We seem to use `endpoint` in most places (e.g. `OTEL_EXPORTER_JAEGER_ENDPOINT` or see receivers/exporters in collector), can this be named `beaconEndpoint` for consistency?

```suggestion
| `beaconEndpoint` ()                         | RUM beacon endpoint, ex. `https://rum-ingest.<realm>.signalfx.com/v1/rum`. If both `realm` and `beaconEndpoint` are set, `beaconEndpoint` takes precedence. [1] [2]                                                        | Yes      |
```

Reviewer 5: +1 for this :)

> | `beaconUrl` ()                         | RUM beacon URL, ex. `https://rum-ingest.<realm>.signalfx.com/v1/rum`. If both `realm` and `beaconUrl` are set, `beaconUrl` takes precedence. [1] [2]                                                        | Yes      |

Reviewer 4: I would propose we mark realm as required and this as not required.

```suggestion
| `beaconUrl` ()                         | RUM beacon URL, ex. `https://rum-ingest.<realm>.signalfx.com/v1/rum`. If both `realm` and `beaconUrl` are set, `beaconUrl` takes precedence. [2]                                                        | No [1]      |
```

Reviewer 5: We could also add a rule that warning should be logged if there is a collision (both are set) - so that user knows which one will be used.

Author: > We could also add a rule that warning should be logged if there is a collision (both are set) - so that user knows which one will be used.

> | `rumAuth` ()                           | RUM authentication token. [1]                                                                                                                                                                               | Yes      |

Reviewer 4: Same question here -- we use the term "access token" for ingest and in UI you can only define these under "access token". Totally agreed that we need to distinguish the RUM token, but perhaps we can make the name consistent.

```suggestion
| `rumAccessToken` ()                           | RUM access token. [1]                                                                                                                                                                               | Yes      |
```

Reviewer 2: `rumAccessToken` is what I had been using in the android configuration, so I in favor of that. :)

Reviewer 2: hrm. no, I was using `rumAuthToken`. Either way, Android would have to change (really not a big deal, though, whatever we decide on).

Reviewer 5: +1 for this as well :)

Author: There was some discussion on this: https://github.com/signalfx/gdi-specification/pull/112#discussion_r657313646

I agree that "access token" is more consistent with UI, but distinguishing RUM tokens from other tokens also makes sense.

> | `applicationName` ()                   | Instrumented application name. [1]                                                                                                                                                                          | Yes      |

Reviewer 4: Continuing on the consistency journey -- we use "service name".

```suggestion
| `serviceName` ()                   | Instrumented service name.                                                                                                                                                                          | Yes  [1]    |
```

Reviewer 5: That's an interesting topic to discuss. Typically "service" is used in a server context. RUM however is client-side. Calling this "service" instead of "application" (which is context-agnostic) may cause some misunderstanding (ie "is this name for a service that ingests the RUM data?").

Author: I think I agree with that, although calling a mobile or a web app a "service" is a bit iffy; also see https://github.com/signalfx/gdi-specification/pull/112#discussion_r657314903

Reviewer 4: Naming is hard -- we technially call it application instrumentation versus serverless instrumentation today -- these refer to server-side. I am not sure if the brower/client is an app -- clearly that term is used on mobile devices.

> | `globalAttributes` ()                  | OpenTelemetry [Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes) that will be added to every span produced by the RUM library. | Yes      |

Reviewer 4: Is this really required?

```suggestion
| `globalAttributes` ()                  | OpenTelemetry [Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes) that will be added to every span produced by the RUM library. | No      |
```

Reviewer 2: This was a requirement of the MVP so yes?

Author: RUM libraries are required to implement that property, yes

| `allowInsecureBeacon` (false)          | Allows sending data to insecure endpoints not using https. [2]                                                                                                                                              | No       |

> | `deploymentEnvironment` ()             | Sets the environment (`deployment.environment` span attribute) for all spans.                                                                                                                               | Yes      |

Reviewer 4: We should not require this -- we can strongly recommend it

```suggestion
| `deploymentEnvironment` ()             | Sets the environment (`deployment.environment` span attribute) for all spans.                                                                                                                               | No      |
```

Author: Hmm, "required" mean "RUM library MUST implement that", though I can see that it is a bit confusing. I'll split the table into two MUST/MAY tables

- [1] Application name, authentication token and either realm or the beacon URL MUST be provided by the user.
  If any of these is missing, the RUM instrumentation library MUST fail to start.
- [2] Implementations MUST enforce by default that `beaconUrl`s are https only, and reject/fail to start otherwise.

>   Implementations MAY offer an `allowInsecureBeacon` option (default false) for customers wishing to use that.

Reviewer 4: suggestion?

```suggestion
  Implementations MAY offer an unsupported `allowInsecureBeacon` option (default false).
```

> Other requirements:

Reviewer 4: What about compression?

Author: Honestly I don't know. Are there any requirements around compression? @johnbley @jkwatson @jtmalinowski

Reviewer 4: My understanding is that we do not compress today -- I am curious why and the cost perf/implications (which I assume are all server-side).

Reviewer 2: There would be additional CPU usage on the client to do compression as well. For Android, we are just doing whatever the zipkin exporter does out of the box, which is whatever the zipkin client does out of the box.

> - RUM library MUST use the Zipkin span exporter by default

Reviewer 4: v1 or v2? JSON or gRPC?

Reviewer 2: V2 JSON

Is there zipkin grpc?

Author: Added v2 JSON to the requirement

> - RUM library MUST limit the number of sent spans to 100 in a 30 second window per `component` attribute value

Reviewer 4: Is this customizable (up or down)? If yes, what happens if you exceed?

Reviewer 2: In the first go-around, it's not configurable, but it would be easy to make it so. Spans get dropped beyond the limit.

Instead, they MUST expose a `SplunkRum` object/class/namespace (depending on the language used) that allows
setting the properties listed below (in a language-specific way: Java may use builders, Swift can use named
parameters with default values, JavaScript can use objects, etc.).

RUM instrumentation libraries MUST support the following configuration properties:

| Property (default value)               | Description                                                                                                                                                                                                 |
| -------------------------------------- | -----------                                                                                                                                                                                                 |
| `realm` ()                             | Splunk realm, e.g. `us0`, `us1`. If set, value of `beaconEndpoint` will be automatically computed based on this. [1] [2] [3]                                                                                |

> | `beaconEndpoint` ()                    | RUM beacon URL, ex. `https://rum-ingest.<realm>.signalfx.com/v1/rum`. If both `realm` and `beaconEndpoint` are set, `beaconEndpoint` takes precedence. [1] [2] [3]                                          |

Reviewer 6: I suggest `e.g.` instead of `ex.` 

AFAIK `ex.` often means "excluding" and usually means "exercise" ;)

| `rumAccessToken` ()                    | RUM authentication token. [1]                                                                                                                                                                               |

> | `serviceName` ()                       | Instrumented application name. [1]                                                                                                                                                                          |

Reviewer 2: I don't think this is a good configuration value. I think it will be very confusing for mobile developers, since the app isn't a "service", and they will probably think this is referring to some backend service that the app might be using. I think we should use `applicationName` or `appName` or `app` instead.

Reviewer 2: (sorry for the late comment..I was going through to implement what was in this PR and ran into this, which had changed since I last looked in detail)

Author: I agree with you and I think that `applicationName` is a much better name - `serviceName` sounds really confusing in the context of RUM.
@flands I think you were the only person that preferred `serviceName`. Can we change that to a less confusing `applicationName`?

Reviewer 7: I'm guessing the ask for this to be `serviceName` was from a desire to be consistent with other parts of the specification. However, given this part of the specification in only for RUM libraries I think it is prudent to ask these kinds of questions and make sure we provide them with names that are meaningful and allow the least amount of cognitive load when they start using our services.

@flands any opposition?

Reviewer 2: Did we really decide to stick with `serviceName`? I think this is a very bad idea, as stated above.

Reviewer 2: ah never mind. Looks like I was looking at at old version. :disappear:

| `globalAttributes` ()                  | OpenTelemetry [Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes) that will be added to every span produced by the RUM library. |

> | `deploymentEnvironment` ()             | Sets the environment (`deployment.environment` span attribute) for all spans.                                                                                                                               |

Reviewer 2: Is there a default value for this? Do we leave it out if the user doesn't provide it?

Author: We leave it out - it's described in the semantic conventions doc
